File: 12-hash/3-collision-resolution/main.py
# Collision Resolution - Python Example for Chaining

class HashMap:
    def __init__(self, size=10):
        self.size = size
        self.buckets = [[] for _ in range(size)]

    def hash(self, key):
        return hash(key) % self.size

    def insert(self, key, value):
        index = self.hash(key)
        for i, (k, v) in enumerate(self.buckets[index]):
            if k == key:
                self.buckets[index][i] = (key, value)
                return
        self.buckets[index].append((key, value))

    def get(self, key):
        index = self.hash(key)
        for k, v in self.buckets[index]:
            if k == key:
                return v
        return None

hmap = HashMap()
hmap.insert("apple", 3)
hmap.insert("banana", 5)
print("Apple count:", hmap.get("apple"))

File: 12-hash/3-collision-resolution/main.cpp
// Collision Resolution - C++ Example (Chaining)

#include <iostream>
#include <vector>
#include <list>
#include <string>

class HashMap {
    std::vector<std::list<std::pair<std::string, int>>> buckets;
    int size;

    int hash(const std::string& key) {
        std::hash<std::string> hasher;
        return hasher(key) % size;
    }

public:
    HashMap(int sz = 10) : size(sz), buckets(sz) {}

    void insert(const std::string& key, int value) {
        int idx = hash(key);
        for (auto& kv : buckets[idx]) {
            if (kv.first == key) {
                kv.second = value;
                return;
            }
        }
        buckets[idx].emplace_back(key, value);
    }

    int get(const std::string& key) {
        int idx = hash(key);
        for (auto& kv : buckets[idx]) {
            if (kv.first == key) return kv.second;
        }
        return -1;  // Not found
    }
};

int main() {
    HashMap hmap;
    hmap.insert("apple", 3);
    hmap.insert("banana", 5);
    std::cout << "Apple count: " << hmap.get("apple") << std::endl;
    return 0;
}
